File: modules/tests/src/commonTest/kotlin/arrays/PassArray.kt
@file:OptIn(ExperimentalUnsignedTypes::class)

package arrays

import natives.test.MyDictionary
import natives.test.MyEnum
import withLib
import kotlin.test.Test
import kotlin.test.assertTrue

class PassArray {

    @Test
    fun passCharArray() = withLib {
        assertTrue(natives.test.passCharArray(charArrayOf('a', 'b')))
    }

    @Test
    fun passCharArrayN() = withLib {
        assertTrue(natives.test.passCharArrayN(null))
    }

    @Test
    fun passBooleanArray() = withLib {
        assertTrue(natives.test.passBooleanArray(booleanArrayOf(true, false)))
    }

    @Test
    fun passByteArray() = withLib {
        assertTrue(natives.test.passByteArray(byteArrayOf(1, 2)))
    }

    @Test
    fun passUByteArray() = withLib {
        assertTrue(natives.test.passUByteArray(ubyteArrayOf(1.toUByte(), UByte.MAX_VALUE)))
    }

    @Test
    fun passShortArray() = withLib {
        assertTrue(natives.test.passShortArray(shortArrayOf(1, 2)))
    }

    @Test
    fun passUShortArray() = withLib {
        assertTrue(natives.test.passUShortArray(ushortArrayOf(1.toUShort(), UShort.MAX_VALUE)))
    }

    @Test
    fun passIntArray() = withLib {
        assertTrue(natives.test.passIntArray(intArrayOf(1, 2)))
    }

    @Test
    fun passUIntArray() = withLib {
        assertTrue(natives.test.passUIntArray(uintArrayOf(1.toUInt(), UInt.MAX_VALUE)))
    }

    @Test
    fun passLongArray() = withLib {
        assertTrue(natives.test.passLongArray(longArrayOf(1, 2)))
    }

    @Test
    fun passULongArray() = withLib {
        assertTrue(natives.test.passULongArray(ulongArrayOf(1.toULong(), ULong.MAX_VALUE)))
    }

    @Test
    fun passFloatArray() = withLib {
        assertTrue(natives.test.passFloatArray(floatArrayOf(1.1f, 2.2f)))
    }

    @Test
    fun passDoubleArray() = withLib {
        assertTrue(natives.test.passDoubleArray(doubleArrayOf(1.1, 2.2)))
    }

    @Test
    fun passStringArray() = withLib {
        assertTrue(natives.test.passStringArray(arrayOf("string1", "string2")))
    }

    @Test
    fun passStringArrayN() = withLib {
        assertTrue(natives.test.passStringArrayN(arrayOf(null, null)))
    }

    @Test
    fun passEnumArray() = withLib {
        assertTrue(natives.test.passEnumArray(arrayOf(MyEnum.CASE1, MyEnum.CASE2)))
    }

    @Test
    fun passDictionaryArray() = withLib {
        assertTrue(natives.test.passDictionaryArray(arrayOf(
            MyDictionary(1, 2, 3, 4),
            MyDictionary(5, 6, 7, 8)
        )))
    }

    @Test
    fun passDictionaryArrayN() = withLib {
        assertTrue(natives.test.passDictionaryArrayN(arrayOf(null, null)))
    }
}

File: modules/tests/natives/testcpp/src/arrays/pass_array.cpp
#include <api.hpp>

bool pass_char_array(KArray<uint16_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 'a' &&
        arg[1] == 'b';
}

bool pass_char_array_n(KOptional<KArray<uint16_t>> arg) {
    return arg.is_none();
}

bool pass_boolean_array(KArray<bool> arg) {
    return arg.get_length() == 2 &&
        arg[0] == true &&
        arg[1] == false;
}

bool pass_byte_array(KArray<int8_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 2;
}

bool pass_ubyte_array(KArray<uint8_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 255u;
}

bool pass_short_array(KArray<int16_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 2;
}

bool pass_ushort_array(KArray<uint16_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 65535u;
}

bool pass_int_array(KArray<int32_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 2;
}

bool pass_uint_array(KArray<uint32_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 4294967295u;
}

bool pass_long_array(KArray<int64_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 2;
}

bool pass_ulong_array(KArray<uint64_t> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1 &&
        arg[1] == 18446744073709551615u;
}

bool pass_float_array(KArray<float> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1.1f &&
        arg[1] == 2.2f;
}

bool pass_double_array(KArray<double> arg) {
    return arg.get_length() == 2 &&
        arg[0] == 1.1 &&
        arg[1] == 2.2;
}

bool pass_string_array(KArray<KString> arg) {
    return arg.get_length() == 2 &&
        std::string(arg[0].get_data(), arg[0].get_size()) == "string1" &&
        std::string(arg[1].get_data(), arg[1].get_size()) == "string2";
}

bool pass_string_array_n(KArray<KOptional<KString>> arg) {
    return arg[0].is_none() && arg[1].is_none();
}

bool pass_enum_array(KArray<MyEnum> arg) {
    return arg.get_length() == 2 &&
        arg[0] == CASE1 &&
        arg[1] == CASE2;
}

bool pass_dictionary_array(KArray<MyDictionary> arg) {
    return arg.get_length() == 2 &&
        arg[0].a == 1 &&
        arg[0].b == 2 &&
        arg[0].c == 3 &&
        arg[0].d == 4 &&
        arg[1].a == 5 &&
        arg[1].b == 6 &&
        arg[1].c == 7 &&
        arg[1].d == 8;
}

bool pass_dictionary_array_n(KArray<KOptional<MyDictionary>> arg) {
    return arg[0].is_none() && arg[1].is_none();
}

bool pass_interface_array(KArray<std::shared_ptr<IMyInterface>> arg) {
    return arg[0]->test() && arg[1]->test();
}

bool pass_interface_array_n(KArray<KOptional<std::shared_ptr<IMyInterface>>> arg) {
    return arg[0].is_none() && arg[1].is_none();
}
